core/wiremessage/msg.go

```diff
@@ -38,12 +64,101 @@ func (m Msg) String() string {
 
 // Len implements the WireMessage interface.
 func (m Msg) Len() int {
-	panic("not implemented")
+	// Header + FlagBits + len of each section + optional checksum
+	totalLen := 16 + 32 // header and flag bits
+
+	for _, section := range m.Sections {
+		totalLen += section.Len()
+	}
+
+	if m.FlagBits&ChecksumPresent > 0 {
+		totalLen += 4
+	}
+
+	return totalLen
 }
 
 // UnmarshalWireMessage implements the Unmarshaler interface.
-func (m *Msg) UnmarshalWireMessage([]byte) error {
-	panic("not implemented")
+func (m *Msg) UnmarshalWireMessage(b []byte) error {
+	var err error
+
+	m.MsgHeader, err = ReadHeader(b, 0)
+	if err != nil {
+		return err
+	}
+	if len(b) < int(m.MsgHeader.MessageLength) {
+		return Error{
+			Type:    ErrOpMsg,
+			Message: "[]byte too small",
+		}
+	}
+
+	m.FlagBits = MsgFlag(readInt32(b, 16))
+
+	// read each section
+	sectionBytes := m.MsgHeader.MessageLength - 16 - 4 // number of bytes taken up by sections
+	hasChecksum := m.FlagBits&ChecksumPresent > 0
+	if hasChecksum {
+		sectionBytes -= 4 // 4 bytes at end for checksum
+	}
+
+	m.Sections = make([]Section, 0)
+	position := 20 // position to read from
+	for sectionBytes > 0 {
+		sectionType := SectionType(b[position])
+		position++
+
+		switch sectionType {
+		case SingleDocument:
+			rdr, size, err := readDocument(b, int32(position))
+			if err.Message != "" {
+				err.Type = ErrOpMsg
+				return err
+			}
+
+			position += size
+			sb := SectionBody{
+				Document: rdr,
+			}
+			sectionBytes -= int32(sb.Len())
+			m.Sections = append(m.Sections, sb)
+		case DocumentSequence:
+			sds := SectionDocumentSequence{}
+			sds.Size = readInt32(b, int32(position))
+			position += 4
+
+			identifier, err := readCString(b, int32(position))
+			if err != nil {
+				return err
+			}
+
+			sds.Identifier = identifier
+			position += len(identifier) + 1 // +1 for \0
+
+			// length of documents to read
+			// sequenceLen - 4 (size) - identifierLength (including \0)
+			docsLen := int(sds.Size) - 4 - len(identifier) - 1
+			for docsLen > 0 {
+				rdr, size, err := readDocument(b, int32(position))
```

Comment:
readDocument here is doing a lot of allocations, one for each document. It would probably be more useful to allocate one large `[]byte` and then break that into the pieces we need. We would need an `appendDocument` function that works like `readDocument` but attempts to append to a given slice and if it can't, grows the slice.

Alternatively, we could just allocate one large slice, copy the bytes, and then make `bson.Reader`s from the slice and do the shallow validation that `readDocument` does.